### ISSUE TITLE
composer install discards local changes in vendors automatically

### ISSUE THREAD

**Issue by EvgenyOrekhov** [composer/composer#4336]
I had a package in vendors that was 1 commit ahead of origin.
I ran `composer install`:
```
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.
```
Then Composer installed that package and discarded my local changes automatically.
Now my commit is lost.
The "discard-changes" option is set to "false".
I expected composer to pause and prompt "You have local changes. Discard them? y/n...".

**Comment by alcohol** [comment]
Your commit is never lost.

**Comment by stof** [comment]
``git reflog`` can give you the hash of the commit which was checkout previously, allowing you to access it again

**Comment by EvgenyOrekhov** [comment]
@stof thanks.
Sorry, just starting to figure out how Composer works.
Still, I didn't expect discarding to be automatical in this case. Is this behavior intentional?

**Comment by ro0NL** [comment]
Perhaps related.. I just lost a commit because I switched from branch in my project (from feature/xxx to master).

The feature branch U was on before switching to master requires an extra composer package, that package was 1 ahead of origin.

After switching my project to master I ran `composer install` and it start removing my extra package installed previous via the feature branch.

My workingcopy got removed.. no reflog :)

I guess it's my own fault and nothing dramatic got lost, but perhaps composer should _always_ give a heads up before it's going to remove stuff which is ahead of origin (commits & modifications)?
